docs/docs/reference/experimental/cc.md

```diff
@@ -0,0 +1,716 @@
+distinction for pure _vs_ impure function values. The distinction is only made in their types.
+
+A closure also captures all capabilities that are captured by the functions
+it calls. For instance, in
+```scala
+def test(fs: FileSystem) =
+  def f() = g()
+  def g() = (x: String) => Logger(fs).log(x)
+  f
+```
+the result of `test` has type `{fs} String => Unit` even though function `f`
+itself does not refer to `fs`.
+
+## Capture Checking of Classes
+
+The principles for capture checking closures also applies to classes. For instance, consider:
+```scala
+class Logger(using fs: FileSystem):
+  def log(s: String): Unit = ... summon[FileSystem] ...
+
+def test(xfs: FileSystem): {xfs} Logger =
+  Logger(xfs)
+```
+Here, class `Logger` retains the capability `fs` as a (private) field. Hence, the result
+of `test` is of type `{xfs} Logger`
+
+Sometimes, a tracked capability is meant to be used only in the constructor of a class, but
+is not intended to be retained as a field. This fact can be communicated to the capture
+checker by declaring the parameter as `@constructorOnly`. Example:
+```scala
+import annotation.constructorOnly
+
+class NullLogger(using @constructorOnly fs: FileSystem):
+  ...
+def test2(using fs: FileSystem): NullLogger = NullLogger() // OK
+```
+
+The captured references of a class include _local capabilities_ and _argument capabilities_. Local capabilities are capabilities defined outside the class and referenced from its body. Argument capabilities are passed as parameters to the primary constructor of the class. Local capabilities are inherited:
+the local capabilities of a superclass are also local capabilities of its subclasses. Example:
+
+```scala
+@capability class Cap
+
+def test(a: Cap, b: Cap, c: Cap) =
+  class Super(y: Cap):
+    def f = a
+  class Sub(x: Cap) extends Super(x)
+    def g = b
+  Sub(c)
+```
+Here class `Super` has local capability `a`, which gets inherited by class
+`Sub` and is combined with `Sub`'s own local capability `b`. Class `Sub` also has an argument capability corresponding to its parameter `x`. This capability gets instantiated to `c` in the final constructor call `Sub(c)`. Hence,
+the capture set of that call is `{a, b, c}`.
+
+The capture set of the type of `this` of a class is inferred by the capture checker, unless the type is explicitly declared with a self type annotation like this one:
+```scala
+class C:
+  self: {a, b} D => ...
+```
+The inference observes the following constraints:
+
+ - The type of `this` of a class `C` includes all captured references of `C`.
+ - The type of `this` of a class `C` is a subtype of the type of `this`
+   of each parent class of `C`.
+ - The type of `this` must observe all constraints where `this` is used.
+
+For instance, in
+```scala
+@capability class Cap
+def test(c: Cap) =
+  class A:
+    val x: A = this
+    def f = println(c)  // error
```

Comment:
I'm not completely clear on this example, if `f` was defined before `x`, would that change the inferred type for `this` and lead to an error when typing `x`?

Reply:
Yes. You get:
```
16 |    val x: A = this
   |               ^^^^
   |               Found:    (A.this : {c} A)
   |               Required: A
```